File: frontend/src/api/applicant.ts
import apiClient from './client'

export interface ApplicantProfile {
  applicant_id: number
  first_name: string
  middle_name: string
  last_name: string
  father_name: string
  gender: string
  dob: string
  caste_id: number
  address: string
  state_id: number
  district_id: number
  pin_code: number
  college_id: number
  other_college?: string | null
  email_id: string
  mobile_no: string
  profile_photo?: string | null
  active_status: string
}

export interface ApplicantCreate {
  first_name: string
  middle_name: string
  last_name: string
  father_name: string
  gender: 'M' | 'F' | 'O'
  dob: string
  caste_id: number
  address: string
  state_id: number
  district_id: number
  pin_code: number
  college_id: number
  other_college?: string
  email_id: string
  mobile_no: string
  profile_photo?: string
}

export interface Application {
  application_id: number
  session_name: string
  center_name: string
  application_status: string
  payment_status: string
  certificate_status: string
  updated_date: string
  reg_id: string | null
}

export interface NewsItem {
  news_id: number
  news_title: string
  news_desc: string
  category_name: string
  start_datetime: string
  end_datetime: string
  status: string
}

export interface Session {
  session_id: number
  session_name: string
  session_desc: string
  start_date: string
  end_date: string
  center_id: number
  active_status: string
}

export interface ApplicationCreate {
  session_id: number
  enroll_id: number;
  qualification_id: number
  stream_id: number
  marks: string
  dob_image?: string
  marksheet_image?: string
  role_id?: number
}

export const applicantAPI = {
  createProfile: async (data: ApplicantCreate): Promise<ApplicantProfile> => {
    const response = await apiClient.post<ApplicantProfile>('/applicant/profile', data)
    return response.data
  },
  getProfile: async (): Promise<ApplicantProfile> => {
    const response = await apiClient.get<ApplicantProfile>('/applicant/profile')
    return response.data
  },
  updateProfile: async (data: Partial<ApplicantCreate>): Promise<ApplicantProfile> => {
    const response = await apiClient.put<ApplicantProfile>('/applicant/profile', data)
    return response.data
  },
  getApplications: async (): Promise<Application[]> => {
    const response = await apiClient.get<Application[]>('/applicant/applications')
    return response.data
  },
  getNews: async (): Promise<NewsItem[]> => {
    const response = await apiClient.get<NewsItem[]>('/applicant/news')
    return response.data
  },
  getSessions: async (): Promise<Session[]> => {
    const response = await apiClient.get<Session[]>('/applicant/sessions')
    return response.data
  },
  createApplication: async (data: ApplicationCreate): Promise<Application> => {
    const response = await apiClient.post<Application>('/applicant/applications', data)
    return response.data
  },
  uploadProfilePhoto: async (file: File): Promise<ApplicantProfile> => {
    const formData = new FormData()
    formData.append('profile_photo', file)
    const response = await apiClient.patch<ApplicantProfile>('/applicant/profile/photo', formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    })
    return response.data
  },
}

File: backend/app/schemas/applicant.py
"""
Applicant schemas
"""

from datetime import date
from typing import Optional
from pydantic import BaseModel, EmailStr, Field


class ApplicantCreate(BaseModel):
    """Schema for creating an applicant profile"""
    first_name: str = Field(..., max_length=30)
    middle_name: str = Field(..., max_length=30)
    last_name: str = Field(..., max_length=30)
    father_name: str = Field(..., max_length=50)
    gender: str = Field(..., pattern="^[MFO]$")  # M, F, or O
    dob: date
    caste_id: int
    address: str
    state_id: int
    district_id: int
    pin_code: int = Field(..., ge=100000, le=999999)
    college_id: int
    other_college: Optional[str] = Field(None, max_length=60)
    email_id: EmailStr
    mobile_no: str = Field(..., pattern="^[0-9]{10}$")
    profile_photo: Optional[str] = None


class ApplicantUpdate(BaseModel):
    """Schema for updating an applicant profile"""
    first_name: Optional[str] = Field(None, max_length=30)
    middle_name: Optional[str] = Field(None, max_length=30)
    last_name: Optional[str] = Field(None, max_length=30)
    father_name: Optional[str] = Field(None, max_length=50)
    gender: Optional[str] = Field(None, pattern="^[MFO]$")
    dob: Optional[date] = None
    caste_id: Optional[int] = None
    address: Optional[str] = None
    state_id: Optional[int] = None
    district_id: Optional[int] = None
    pin_code: Optional[int] = Field(None, ge=100000, le=999999)
    college_id: Optional[int] = None
    other_college: Optional[str] = Field(None, max_length=60)
    mobile_no: Optional[str] = Field(None, pattern="^[0-9]{10}$")
    profile_photo: Optional[str] = None


class ApplicantResponse(BaseModel):
    """Schema for applicant profile response"""
    applicant_id: int
    first_name: str
    middle_name: str
    last_name: str
    father_name: str
    gender: str
    dob: date
    caste_id: int
    address: str
    state_id: int
    district_id: int
    pin_code: int
    college_id: int
    other_college: Optional[str]
    email_id: str
    mobile_no: str
    profile_photo: Optional[str]
    active_status: str
    
    class Config:
        from_attributes = True

